server/src/main/java/com/paypal/selion/proxy/SeLionRemoteProxy.java

```diff
@@ -245,15 +245,28 @@ public synchronized void shutdownNode() {
 
         @Override
         public void run() {
+            int timeout = Integer.MAX_VALUE; // Default. Wait a very long time.
+            int expired = 0;
+            try {
+                timeout = ConfigParser.parse().getInt("nodeRecycleThreadWaitTimeout");
+            } catch (ConfigParserException e) { // NOSONAR
+                // ignore we have a default already
+            }
             appendMsgToCustomLog("Started NodeRecycleThread to restart the node once all the sessions are completed.");
-            while (getTotalUsed() > 0) {
+            while ((getTotalUsed() > 0) && (expired < timeout)) {
```

Comment:
Yes. You can create a test that polls the browser and keeps it open indefinitely and run some "well-behaved" tests which reach the `uniqueSessionCount` and trigger the `NodeRecycleThread` ... In this scenario, the `NodeRecycleThread` will keep looping as long as that long running "mis-behaved" test keeps polling